# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [1]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name,Transported
0,0001_01,Europa,False,B/0/P,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False
1,0002_01,Earth,False,F/0/S,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True
2,0003_01,Europa,False,A/0/S,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False
3,0003_02,Europa,False,A/0/S,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False
4,0004_01,Earth,False,F/1/S,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True


Now perform the same as before:
- Feature Scaling
- Feature Selection


In [4]:
def df_info(df):
    print("\n    ---- df info ---")
    print(df.info())
    print("\n    ---- Empty Values ---")
    print(df.isna().sum())
    print("\n    ----- Dtypes -----")
    print(df.dtypes)
    print("\n    ---- Statistics ---")
    return df.describe()
df_info(spaceship)


    ---- df info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB
None

    ---- Empty Values ---
PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destinat

,Age,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck
count,8514.000000,8512.000000,8510.000000,8485.000000,8510.000000,8505.000000
mean,28.827930,224.687617,458.077203,173.729169,311.138778,304.854791
std,14.489021,666.717663,1611.489240,604.696458,1136.705535,1145.717189
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,19.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,27.000000,0.000000,0.000000,0.000000,0.000000,0.000000
75%,38.000000,47.000000,76.000000,27.000000,59.000000,46.000000
max,79.000000,14327.000000,29813.000000,23492.000000,22408.000000,24133.000000


In [15]:
X=spaceship

In [17]:
X.rename(columns={col : col.strip() for col in X.columns}, inplace = True)
X.rename(columns={col : col.replace(" ", "_").lower() for col in X.columns}, inplace = True)
X ["cabin"] = X ["cabin"].str[0]
X = X.dropna()
X.head()

,passengerid,homeplanet,cryosleep,cabin,destination,age,vip,roomservice,foodcourt,shoppingmall,spa,vrdeck,name,transported
0,0001_01,Europa,False,B,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False
1,0002_01,Earth,False,F,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True
2,0003_01,Europa,False,A,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False
3,0003_02,Europa,False,A,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False
4,0004_01,Earth,False,F,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True


In [21]:
target = X["transported"]
features = X.drop(columns = ["passengerid","name", "transported"])
features = pd.get_dummies (features, drop_first = True)

**Perform Train Test Split**

In [31]:
X_train,X_test, y_train,y_test = train_test_split(features,target,test_size=0.2,random_state=0)
X_train.head(5)

,age,roomservice,foodcourt,shoppingmall,spa,vrdeck,homeplanet_Europa,homeplanet_Mars,cryosleep_True,cabin_B,cabin_C,cabin_D,cabin_E,cabin_F,cabin_G,cabin_T,destination_PSO J318.5-22,destination_TRAPPIST-1e,vip_True
3432,32.0,0.0,0.0,0.0,0.0,0.0,False,True,True,False,False,False,True,False,False,False,False,True,False
7312,4.0,0.0,0.0,0.0,0.0,0.0,False,False,True,False,False,False,False,False,True,False,False,True,False
2042,30.0,0.0,236.0,0.0,1149.0,0.0,False,False,False,False,False,False,False,True,False,False,False,True,False
4999,17.0,13.0,0.0,565.0,367.0,1.0,False,True,False,False,False,False,True,False,False,False,False,True,False
5755,26.0,0.0,0.0,0.0,0.0,0.0,False,False,True,False,False,False,False,False,True,False,False,True,False


In [32]:
from sklearn.preprocessing import MinMaxScaler

In [33]:
scaler = MinMaxScaler()

X_train_norm = pd.DataFrame(scaler.fit_transform(X_train), columns=X_train.columns)

X_test_norm = pd.DataFrame(scaler.transform(X_test), columns=X_test.columns)

In [36]:
display(X_train_norm.head(5))
display(X_test_norm.head(5))

,age,roomservice,foodcourt,shoppingmall,spa,vrdeck,homeplanet_Europa,homeplanet_Mars,cryosleep_True,cabin_B,cabin_C,cabin_D,cabin_E,cabin_F,cabin_G,cabin_T,destination_PSO J318.5-22,destination_TRAPPIST-1e,vip_True
0,0.405063,0.00000,0.000000,0.000000,0.000000,0.000000,0.0,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
1,0.050633,0.00000,0.000000,0.000000,0.000000,0.000000,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0
2,0.379747,0.00000,0.007916,0.000000,0.051276,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0
3,0.215190,0.00131,0.000000,0.046111,0.016378,0.000049,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
4,0.329114,0.00000,0.000000,0.000000,0.000000,0.000000,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0


,age,roomservice,foodcourt,shoppingmall,spa,vrdeck,homeplanet_Europa,homeplanet_Mars,cryosleep_True,cabin_B,cabin_C,cabin_D,cabin_E,cabin_F,cabin_G,cabin_T,destination_PSO J318.5-22,destination_TRAPPIST-1e,vip_True
0,0.632911,0.0,0.000000,0.0,0.00000,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0
1,0.227848,0.0,0.000000,0.0,0.00000,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0
2,0.189873,0.0,0.000000,0.0,0.00000,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
3,0.658228,0.0,0.000000,0.0,0.00000,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0
4,0.784810,0.0,0.054775,0.0,0.07774,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0


**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

- Bagging and Pasting

In [37]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, root_mean_squared_error

In [61]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.metrics import f1_score, precision_score, recall_score

PARAMS_BAGGING = {
    'n_estimators': 100,
    'max_samples': 1000,
    'random_state': 42
}

PARAMS_TREE = {
    'max_depth': 20,
    'random_state': 42
}

# --- TRAINING ---
base_tree = DecisionTreeClassifier(**PARAMS_TREE)
bagging_clf = BaggingClassifier(estimator=base_tree, **PARAMS_BAGGING)
bagging_clf.fit(X_train_norm, y_train)

# --- PREDICTIONS AND METRICS ---
pred = bagging_clf.predict(X_test_norm)

print("Accuracy", bagging_clf.score(X_test_norm, y_test))
print("F1 score", f1_score(y_test, pred))
print("Precision", precision_score(y_test, pred))
print("Recall", recall_score(y_test, pred))



Accuracy 0.783661119515885
F1 score 0.7843137254901961
Precision 0.7819548872180451
Recall 0.7866868381240545


In [62]:
# Pasting: sampling without replacement
bagging_pasting = BaggingClassifier(estimator=base_tree, bootstrap=False, **PARAMS_BAGGING)
bagging_pasting.fit(X_train_norm, y_train)
pred_pasting = bagging_pasting.predict(X_test_norm)

print("Accuracy", bagging_pasting.score(X_test_norm, y_test))
print("F1 score", f1_score(y_test, pred_pasting))

Accuracy 0.7829046898638427
F1 score 0.783068783068783


- Random Forests

In [51]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, precision_score, recall_score

forest = RandomForestClassifier()
forest.fit(X_train_norm, y_train)
pred = forest.predict(X_test_norm)

#print("MAE", mean_absolute_error(pred, y_test))
#print("RMSE", np.sqrt(mean_squared_error(pred, y_test)))
print("Accuracy", forest.score(X_test_norm, y_test))
print("F1 score", f1_score(y_test, pred))
print("Precision", precision_score(y_test, pred))
print("Recall", recall_score(y_test, pred))

Accuracy 0.7912254160363086
F1 score 0.7896341463414634
Precision 0.7956989247311828
Recall 0.783661119515885


- Gradient Boosting

In [54]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import f1_score, precision_score, recall_score

gb_classifier = GradientBoostingClassifier()
gb_classifier.fit(X_train_norm, y_train)
pred = gb_classifier.predict(X_test_norm)

print("Accuracy", gb_classifier.score(X_test_norm, y_test))
print("F1 score", f1_score(y_test, pred))
print("Precision", precision_score(y_test, pred))
print("Recall", recall_score(y_test, pred))

Accuracy 0.7866868381240545
F1 score 0.7959479015918958
Precision 0.7628294036061026
Recall 0.8320726172465961


- Adaptive Boosting

In [59]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score, precision_score, recall_score

adapt_boost = AdaBoostClassifier(DecisionTreeClassifier())
adapt_boost.fit(X_train_norm, y_train)
pred = adapt_boost.predict(X_test_norm)  

print("Accuracy", adapt_boost.score(X_test_norm, y_test))
print("F1 score", f1_score(y_test, pred))
print("Precision", precision_score(y_test, pred))
print("Recall", recall_score(y_test, pred))

Accuracy 0.754160363086233
F1 score 0.7660187185025198
Precision 0.7307692307692307
Recall 0.8048411497730711


Which model is the best and why?

results = pd.DataFrame({
    'Model':     ['Random Forest', 'Gradient Boosting', 'AdaBoost'],
    'Accuracy':  [0.7912, 0.7867, 0.7542],
    'F1':        [0.7896, 0.7959, 0.7660],
    'Precision': [0.7957, 0.7628, 0.7308],
    'Recall':    [0.7837, 0.8321, 0.8048]
})

results

Random Forest is the best model overall.

It has the highest accuracy and precision, and its F1 score is almost 
as good as Gradient Boosting's. So it makes fewer mistakes overall and 
is more balanced than the other two.

Gradient Boosting has the best recall, so it catches more positive cases, 
but it also makes more false positives (lower precision).

AdaBoost is the weakest model in every metric.

Overall, Random Forest seems like the safest and most balanced choice.